Code Kingdoms JavaScript Editor

inversion opened the pull request:
This is to add the Code Kingdoms JavaScript editor to TouchDevelop.

I have added some comments in the diff at places which need review.

msftclas commented:
Hi __@inversion__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

inversion commented on editor/external.ts:
```diff
@@ -32,7 +32,19 @@ module TDev {
+          id: 'codekingdoms',
+          origin: origin,
+          path: path + 'codekingdoms/editor.html'
+          // Local testing
```
do we leave this in? is there a better way for us to get this working locally with the same code?

inversion commented on www/codekingdoms/codekingdoms-main.ts:
```diff
@@ -0,0 +1,484 @@
+  }
+
+  function loadCK(s: string) {
+    // var text = s || "<xml></xml>";
```
review

inversion commented on www/codekingdoms/codekingdoms-main.ts:
```diff
@@ -0,0 +1,484 @@
+  }
+
+  function saveCK(): string {
+    // var xml = Blockly.Xml.workspaceToDom(Blockly.mainWorkspace);
```
review

inversion commented on www/codekingdoms/codekingdoms-main.ts:
```diff
@@ -0,0 +1,484 @@
+    dirty = false;
+  }
+
+  /*function compileOrError(msgSel?: string) {
```
review dead code

inversion commented on www/codekingdoms/codekingdoms-main.ts:
```diff
@@ -0,0 +1,484 @@
+    if (fota)
+      $("body").addClass("hasFota");
+
+    // $("#command-quit").click(() => {
```
review dead code

inversion commented on www/codekingdoms/editor.html:
```diff
@@ -0,0 +1,65 @@
+      </div>
+      <div class="hbox flex1">
+          <div id="editor" class="flex3">
+          		<iframe id="ck_frame" src="http://localhost:8888/ck-client/game/">
```
hardcoded localhost

inversion commented on www/codekingdoms/style.css:
```diff
@@ -0,0 +1,272 @@
+/* Re-usable, generic rules */
```
Are all these styles needed or is there duplication from Blockly's CSS?

inversion commented:
Also noticed the contributing guidelines do not allow merge commits so we can rebase this after fixing the above issues.

pelikhan commented on editor/scriptList.ts:
```diff
@@ -6145,6 +6145,7 @@
 
         static editorIcons = {
             "blockly": "blockeditor,#517EC0",
+            "codekingdoms": "fa-codekingdoms,#fbf299",
```
is fa-codekingdoms defined in svg.ts?

inversion replied:
@pelikhan Fixing that now, with regards to the other logo at logoUrl in external.ts I see the format for the Blockly logo at ```https://microbit0.blob.core.windows.net/pub/vtadprog``` is a HTML page containing a PNG. Is this the required format or can we link directly to the PNG?

inversion replied:
@pelikhan Can you please explain the format for embedding a logo in icons:any in svg.ts? It seems to just be data for one ```<path>```. Can we not embed an arbitrary SVG as our icon?

pelikhan replied:
You can embed your own svg and I will fix it on our end.

pelikhan replied:
you could use a data-uri with svg embedded.

inversion commented:
@pelikhan I've added our SVG logo (small square logo for the editor selection popup) to svg.ts but it's not referenced currently. I have also squashed the commits into one and got rid of the nasty merges.

Thanks for your comments, please let me know what else we need to continue.

pelikhan commented:
You still need to go through the bot instructions before I can pull this in...
```
In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
```

inversion commented:
@pelikhan Done, thanks.

msftclas commented:
__@inversion__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;
